shaping/render_test.go

```diff
@@ -0,0 +1,182 @@
+		}
+
+		last := points[len(points)-1]
+
+		if seg.Op == api.SegmentOpLineTo {
+			for t := float32(0); t < 1; t += 0.2 {
+				middleX := int((t*current.X + (1-t)*last.X) * factor)
+				middleY := -int((t*current.Y + (1-t)*last.Y) * factor)
+				drawPoint(img, dot.Add(image.Pt(middleX, middleY)), c)
+			}
+		}
+
+		current = last
+	}
+}
+
+var (
+	red   = color.RGBA{R: 0xFF, A: 0xFF}
+	green = color.RGBA{R: 0xCF, G: 0xFF, B: 0xCF, A: 0xFF}
+	black = color.RGBA{A: 0xFF}
+)
+
+func imageDims(line []Output) (width, height, baseline int) {
+	firstRun := line[0]
+	if firstRun.Direction.IsVertical() {
+		baseline = -firstRun.GlyphBounds.Descent.Round()
+		for _, run := range line {
+			if w := run.GlyphBounds.LineThickness().Round(); w > width {
+				width = w
+			}
+			height += -run.Advance.Round()
+		}
+	} else {
+		baseline = firstRun.GlyphBounds.Ascent.Round()
+		for _, run := range line {
+			if h := run.GlyphBounds.LineThickness().Round(); h > height {
+				height = h
+			}
+			width += run.Advance.Round()
+		}
+	}
+	return
+}
+
+func drawTextLine(runs []Output, file string) error {
+	width, height, baseline := imageDims(runs)
+	img := image.NewRGBA(image.Rect(0, 0, width, height))
+	// white background
+	draw.Draw(img, img.Rect, image.NewUniform(color.White), image.Point{}, draw.Src)
+
+	if runs[0].Direction.IsVertical() {
+		// draw the baseline
+		drawVLine(img, image.Pt(baseline, 0), height, black)
+
+		dot := image.Pt(baseline, 0)
+		for _, run := range runs {
+			dot = drawVRun(run, img, dot)
+		}
+	} else {
+		// draw the baseline
+		drawHLine(img, image.Pt(0, baseline), width, black)
+
+		dot := image.Pt(0, baseline)
+		for _, run := range runs {
+			dot = drawHRun(run, img, dot)
+		}
+	}
+
+	f, err := os.Create(file)
+	if err != nil {
+		return err
+	}
+	if err = png.Encode(f, img); err != nil {
+		return err
+	}
+	err = f.Close()
+	return err
+}
+
+// assume horizontal direction
+func drawHRun(out Output, img *image.RGBA, dot image.Point) image.Point {
+	sizeFactor := float32(out.Size.Round()) / float32(out.Face.Upem())
+	for _, g := range out.Glyphs {
+		minX := dot.X + g.XOffset.Round() + g.XBearing.Round()
+		maxX := minX + g.Width.Round()
+		minY := dot.Y + g.YOffset.Round() - g.YBearing.Round()
+		maxY := minY - g.Height.Round()
+
+		drawRect(img, image.Pt(minX, minY), image.Pt(maxX, maxY), green)
+
+		// draw the dot
+		drawPoint(img, dot, black)
+
+		// draw a sketch of the glyphs
+		glyphData := out.Face.GlyphData(g.GlyphID).(api.GlyphOutline)
+		drawGlyph(img, dot.Add(image.Pt(g.XOffset.Ceil(), g.YOffset.Ceil())), glyphData, sizeFactor, black)
+
+		dot.X += g.XAdvance.Round()
+		// draw the advance
+		drawVLine(img, image.Pt(dot.X, 0), img.Bounds().Dy(), red)
+	}
+
+	return dot
+}
+
+// assume vertical direction
+func drawVRun(out Output, img *image.RGBA, dot image.Point) image.Point {
+	sizeFactor := float32(out.Size.Round()) / float32(out.Face.Upem())
+	for _, g := range out.Glyphs {
+		dot.Y += -g.YAdvance.Round()
```

Comment:
Question: should Y advance earlier than drawPoints?

Reply:
Actually I think both ways are OK. As it is, the first glyph is draw under the horizontal line Y=0, so that it is in the image bounds. 
In your case, it is perhaps better to display the first glyph over the line.

Reply:
> Actually I think both ways are OK.

YAdvance values are different among Japanese glyphs and English glyphs, so I don't understand why both could be ok.

Reply:
Well, if the advances are y1, y2, ... yn the dot is either at 
0, y1, y1+y2, y1+y2+y3, ... 
or at 
y1, y1+y2, y1+y2+y3, ...
so that the whole run is the same, but shifted by y1

Reply:
When rendering "あaあ", and "あ" has 10 y advance and "a" has 5 y advance,

1. あ: 0, a: 10, あ: 15
2. あ: 10, a: 15, あ: 25

And in the case of 1, "a" and the last "あ" can be overlapped.

Reply:
I do not understand why scenario 1 is a problem. That seems correct to me. You draw the glyph before shifting the dot by the advance. Is the glyph for `a` bigger than its Y advance? Could you share a screenshot of the issue?

Reply:
1. Apply this patch onto this PR
```diff
diff --git a/shaping/render_test.go b/shaping/render_test.go
index 6b056f9..e47cd3e 100644
--- a/shaping/render_test.go
+++ b/shaping/render_test.go
@@ -154,8 +154,6 @@ func drawHRun(out Output, img *image.RGBA, dot image.Point) image.Point {
 func drawVRun(out Output, img *image.RGBA, dot image.Point) image.Point {
        sizeFactor := float32(out.Size.Round()) / float32(out.Face.Upem())
        for _, g := range out.Glyphs {
-               dot.Y += -g.YAdvance.Round()
-
                minX := dot.X + g.XOffset.Round() + g.XBearing.Round()
                maxX := minX + g.Width.Round()
 
@@ -176,6 +174,8 @@ func drawVRun(out Output, img *image.RGBA, dot image.Point) image.Point {
 
                // draw the advance
                drawHLine(img, image.Pt(0, dot.Y), img.Bounds().Dx(), red)
+
+               dot.Y += -g.YAdvance.Round()
        }
 
        return dot
```

2. Run `go test ./shaping/ -run=TestShapeVerticalScripts`
3. Open `/tmp/shape_vert_mixed.png`

![image](https://github.com/go-text/typesetting/assets/16950/1390b66b-e676-4918-8193-ec786f44874d)